From f7512703bc119072ccc205479d96c047f10d4812 Mon Sep 17 00:00:00 2001
From: lieppos <l@l>
Date: Mon, 5 Oct 2026 00:54:04 +0300
Subject: [PATCH 2/2] rpmb-mtk: GQ5012BF1 TrustKernel raw-frame RPMB ioctls
 (10..14), multi-block defines, netlink off, raw-frame helpers built without
 Trustonic - 17/17 functions, all sections except build-id

---
 drivers/char/rpmb/rpmb-mtk.c | 135 +++++++++++++++++++++++++++++++++--
 1 file changed, 129 insertions(+), 6 deletions(-)

diff --git a/drivers/char/rpmb/rpmb-mtk.c b/drivers/char/rpmb/rpmb-mtk.c
index 318bd98..b3cc32e 100644
--- a/drivers/char/rpmb/rpmb-mtk.c
+++ b/drivers/char/rpmb/rpmb-mtk.c
@@ -57,7 +57,7 @@ static struct mmc_host *mtk_mmc_host[] = {NULL};
 
 /* #define __RPMB_MTK_DEBUG_MSG */
 /* #define __RPMB_MTK_DEBUG_HMAC_VERIFY */
-#define __RPMB_KERNEL_NL_SUPPORT
+/* #define __RPMB_KERNEL_NL_SUPPORT */
 #define __RPMB_IOCTL_SUPPORT
 
 /* TEE usage */
@@ -132,6 +132,35 @@ static struct nl_rpmb_send_req nl_rpmb_req;
 #define MAX_RPMB_TRANSFER_BLK (1U)
 #endif
 
+/*
+ * GQ5012BF1 (TrustKernel TEE, no Trustonic): multi-block RPMB access is
+ * enabled without drrpmb_Api.h (proven by the eMMC multi-block code paths
+ * and the 16-block transfer limit in the stock module).
+ */
+#ifndef RPMB_MULTI_BLOCK_ACCESS
+#define RPMB_MULTI_BLOCK_ACCESS 1
+#undef MAX_RPMB_TRANSFER_BLK
+#define MAX_RPMB_TRANSFER_BLK (16U)
+#define MAX_RPMB_REQUEST_SIZE (512U*MAX_RPMB_TRANSFER_BLK)
+#endif
+
+/* GQ5012BF1: raw-frame RPMB ioctls used by TrustKernel teed on /dev/rpmb0 */
+#define RPMB_IOCTL_TK_WRITE_DATA	10
+#define RPMB_IOCTL_TK_READ_DATA		11
+#define RPMB_IOCTL_TK_GET_CNT		12
+#define RPMB_IOCTL_TK_GET_WR_SIZE	13
+#define RPMB_IOCTL_TK_PROGRAM_KEY	14
+#define RPMB_TK_FRAME_SIZE		512
+#define RPMB_TK_MAX_BUF_SIZE		4096
+
+u16 rpmb_get_rw_size(struct rpmb_dev *rdev);
+
+struct rpmb_tk_ioc_param {
+	u32 blk_cnt;
+	u32 rw_size;
+	u8 __user *frames;
+};
+
 #define RPMB_NAME "rpmb"
 
 #define MIN(a, b) (((a) < (b)) ? (a) : (b))
@@ -669,7 +698,6 @@ static int rpmb_req_get_wc_ufs(u8 region, u8 *keybytes, u32 *wc, u8 *frame)
 	return ret;
 }
 
-#if IS_ENABLED(CONFIG_TRUSTONIC_TEE_SUPPORT)
 static int rpmb_req_read_data_ufs(u8 *frame, u32 blk_cnt)
 {
 	struct rpmb_data rpmbdata;
@@ -784,6 +812,7 @@ static int rpmb_req_write_data_ufs(u8 *frame, u32 blk_cnt)
 	return ret;
 }
 
+#if IS_ENABLED(CONFIG_TRUSTONIC_TEE_SUPPORT)
 static int rpmb_req_purge_enable(u8 region, u8 *frame)
 {
 	int ret = 0;
@@ -867,6 +896,7 @@ static int rpmb_req_read_purge_status(u8 region, u8 *frame)
 	return ret;
 }
 
+#endif
 
 static int rpmb_req_program_key_ufs(u8 region, u8 *frame, u32 blk_cnt)
 {
@@ -922,7 +952,6 @@ static int rpmb_req_program_key_ufs(u8 region, u8 *frame, u32 blk_cnt)
 
 	return ret;
 }
-#endif
 
 static int rpmb_req_ioctl_write_data_ufs(struct rpmb_ioc_param *param)
 {
@@ -1808,6 +1837,7 @@ int emmc_rpmb_req_set_key(struct mmc_card *card, u8 *key)
 	return ret;
 }
 
+#if IS_ENABLED(CONFIG_TRUSTONIC_TEE_SUPPORT)
 static int rpmb_gp_execute_emmc(u32 cmdId)
 {
 	int ret;
@@ -1890,6 +1920,7 @@ static int rpmb_gp_execute_emmc(u32 cmdId)
 
 	return 0;
 }
+#endif
 
 int rpmb_req_get_wc_emmc(struct mmc_card *card, u8 *key, u32 *wc)
 {
@@ -2622,14 +2653,37 @@ static long rpmb_ioctl_ufs(struct file *pfile, unsigned int cmd, unsigned long a
 	int err = 0;
 	unsigned long n;
 	struct rpmb_ioc_param param;
+	struct rpmb_tk_ioc_param tk_param;
+	struct rpmb_dev *rawdev_ufs_rpmb;
+	u8 *frame = NULL;
 
-	n = copy_from_user(&param, (void *)arg, sizeof(param));
-
+	n = copy_from_user(&tk_param, (void *)arg, sizeof(tk_param));
 	if (n != 0UL) {
-		MSG(ERR, "%s, copy from user failed: %x\n", __func__, err);
+		MSG(ERR, "%s, copy param from user failed\n", __func__);
 		return -EFAULT;
 	}
 
+	if (tk_param.blk_cnt * RPMB_TK_FRAME_SIZE > RPMB_TK_MAX_BUF_SIZE) {
+		MSG(ERR, "%s, rpmb buffer too large\n", __func__);
+		return -EINVAL;
+	}
+
+	if (tk_param.blk_cnt) {
+		frame = kmalloc(tk_param.blk_cnt * RPMB_TK_FRAME_SIZE, GFP_KERNEL);
+		if (!frame) {
+			MSG(ERR, "%s, failed to malloc rpmb buffer\n", __func__);
+			return -ENOMEM;
+		}
+
+		n = copy_from_user(frame, tk_param.frames,
+				   tk_param.blk_cnt * RPMB_TK_FRAME_SIZE);
+		if (n != 0UL) {
+			MSG(ERR, "%s, copy frame from user failed\n", __func__);
+			kfree(frame);
+			return -EFAULT;
+		}
+	}
+
 	switch (cmd) {
 
 	case RPMB_IOCTL_PROGRAM_KEY:
@@ -2674,6 +2728,75 @@ static long rpmb_ioctl_ufs(struct file *pfile, unsigned int cmd, unsigned long a
 
 		break;
 
+	case RPMB_IOCTL_TK_READ_DATA:
+		err = rpmb_req_read_data_ufs(frame, tk_param.blk_cnt);
+		if (err) {
+			MSG(ERR, "%s: rpmb read request IO error(%x)\n", __func__, err);
+			kfree(frame);
+			return err;
+		}
+		err = copy_to_user(tk_param.frames, frame,
+				   tk_param.blk_cnt * RPMB_TK_FRAME_SIZE);
+		kfree(frame);
+		if (err)
+			MSG(ERR, "%s: copy to user failed: %x\n", __func__, err);
+		break;
+
+	case RPMB_IOCTL_TK_WRITE_DATA:
+		err = rpmb_req_write_data_ufs(frame, tk_param.blk_cnt);
+		if (err) {
+			MSG(ERR, "%s: rpmb write request IO error(%x)\n", __func__, err);
+			kfree(frame);
+			return err;
+		}
+		err = copy_to_user(tk_param.frames, frame, RPMB_SZ_FRAME);
+		kfree(frame);
+		if (err)
+			MSG(ERR, "%s: copy to user failed: %x\n", __func__, err);
+		break;
+
+	case RPMB_IOCTL_TK_GET_CNT:
+		err = rpmb_req_get_wc_ufs(RPMB_REGION0, NULL, NULL, frame);
+		if (err) {
+			MSG(ERR, "%s: rpmb get write counter error(%x)\n", __func__, err);
+			kfree(frame);
+			return err;
+		}
+		err = copy_to_user(tk_param.frames, frame,
+				   tk_param.blk_cnt * RPMB_TK_FRAME_SIZE);
+		kfree(frame);
+		if (err)
+			MSG(ERR, "%s: copy to user failed: %x\n", __func__, err);
+		break;
+
+	case RPMB_IOCTL_TK_GET_WR_SIZE:
+		rawdev_ufs_rpmb = ufs_mtk_rpmb_get_raw_dev();
+		if (rawdev_ufs_rpmb) {
+			tk_param.rw_size = rpmb_get_rw_size(rawdev_ufs_rpmb);
+			err = copy_to_user((void *)arg, &tk_param, sizeof(tk_param));
+			if (err)
+				MSG(ERR, "%s: copy to user failed: %x\n", __func__, err);
+		} else
+			err = RPMB_ALLOC_ERROR;
+		break;
+
+	case RPMB_IOCTL_TK_PROGRAM_KEY:
+		if (tk_param.blk_cnt != 1) {
+			kfree(frame);
+			return -EINVAL;
+		}
+		err = rpmb_req_program_key_ufs(RPMB_REGION0, frame, tk_param.blk_cnt);
+		if (err) {
+			MSG(ERR, "%s: program key failed: %x\n", __func__, err);
+			kfree(frame);
+			return err;
+		}
+		err = copy_to_user(tk_param.frames, frame, RPMB_SZ_FRAME);
+		kfree(frame);
+		if (err)
+			MSG(ERR, "%s: copy to user failed: %x\n", __func__, err);
+		break;
+
 	default:
 		MSG(ERR, "%s, wrong ioctl code (%d)!!!\n", __func__, cmd);
 		err = -ENOTTY;
-- 
2.56.0

